Dcc loco commands sender

sirmax123 opened the pull request.

sirmax123 committed:
Simple class for sending commnads to loco

sirmax123 committed:
Fix typo in emergency packet handler

sirmax123 committed:
Simple examples for loco control and command station

sirmax123 committed:
fix typo

surabujin commented on dccpi_mm/dcc_loco_control.py:
```diff
@@ -0,0 +1,98 @@
+        """
+        Simple key handler
+        """
+        if key in self.UP_SPEED_KEYS:
```
increase/decrease is more intuitive terms here (as for me).

sirmax123 replied:
Fixed

surabujin commented on dccpi_mm/dcc_loco_control.py:
```diff
@@ -0,0 +1,98 @@
+            # Set direction if not defined.
+            # Direction can be changed only if speed = 0
+            if ( self.loco_speed == 0 ) and ( self.loco_direction != 'forward' ):
+                self.loco_direction = 'forward'
```
I would not keep "direction" marker, if it depends only from "speed" value (above or belove 0) (you can make calculated property (@property) for easy access to this value).

sirmax123 replied:
> I would not keep "direction" marker, if it depends only from "speed" value (above or belove 0) (you can make calculated property (@Property) for easy access to this value).

It can't be calculated from speed.
"Forward Lamp" function will switch on/off lamp depends on direction even if speed is Zero.  For 'forward' direction it is a headlight and for 'reverse' it is taillights.

surabujin commented on dccpi_mm/dcc_loco_control.py:
```diff
@@ -0,0 +1,98 @@
+            if ( self.loco_speed == 0 ) and ( self.loco_direction != 'forward' ):
+                self.loco_direction = 'forward'
+            # Max speed is 14
+            elif (self.loco_speed < 14):
```
easier way:
`self.loco_speed = min(self.loco_speed + 1, 14)`

sirmax123 replied:
Fixed

surabujin commented on dccpi_mm/dcc_loco_control.py:
```diff
@@ -0,0 +1,98 @@
+            'speed':        abs(self.loco_speed)
+        }
+
+        # TODO: Move function keys to predef. constatn? not sure.
```
ep - constants are much more readable

sirmax123 replied:
Fixed

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -58,7 +58,7 @@ def main_loop(self):
         while True:
```
"lock" this "priority" logic into some "CommandReader" class (interface can be iterrartor/generator or callable object or just .get_next()). It will be easy to use and easy to extend/replace.

PS Same for producer/writer.

surabujin commented on dccpi_mm/dcc_loco_control.py:
```diff
@@ -0,0 +1,98 @@
+            if key in ('0'):
+                # FL (Forward Lamp) is spetial case.
+                # (value + 1) % 2 inverts function value on each step
+                self.loco_functions['FL'] = (self.loco_functions['FL'] + 1) % 2
```
it can be boolean (something like `value = not value`)?

surabujin commented on dccpi_mm/dcc_loco_control.py:
```diff
@@ -0,0 +1,98 @@
+            else:
+                self.loco_functions[key] = (self.loco_functions[key] + 1) % 2
+
+        command_function = {
```
you need to produce 2 kind of commands despite key pressed? I mean that you produce commands even it no one key "functions" have been matched.

sirmax123 replied:
Fixed

sirmax123 committed:
PEP8: fixed warnings

sirmax123 committed:
PEP8: fixed warnings

sirmax123 committed:
Add configurable idle packets number

sirmax123 committed:
Fix logger import: python3 compatibility

sirmax123 committed:
Add python3 support for dcc_packet_factory.py

sirmax123 committed:
Python3 support to import

sirmax123 committed:
Modify variables name to make them more human-readble

sirmax123 committed:
Make code more readble: min/max instead of if/else

sirmax123 committed:
Refactoring

sirmax123 committed:
PEP8

sirmax123 committed:
PEP8 __init__.py

sirmax123 committed:
fix typo

sirmax123 committed:
Fix speed for reverse mode

sirmax123 committed:
Add support for bool in function controls

sirmax123 committed:
Refactoring

sirmax123 committed:
add dependencies support, increase version number

sirmax123 committed:
- Refactored queue handler
- Added queue cleaner during startup.

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,114 @@
         self.e_stop_packet = self.packet_factory.DCCEStopPacket().to_bit_string()
-        self.commands_queue = RedisQueue(commands_queue, **redis_kwargs)
-        self.emergency_queue = RedisQueue(emergency_queue, **redis_kwargs)
+        self.commands_queue = QueueReader(commands_queue, emergency_queue, **redis_kwargs)
```
Ты можешь передать ответвенность за создание `QueueReader' в вызывающую сторону. Т.е. тут ты получаешь уже готовый объект и пользуешь его для чтения команд.

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,114 @@
         """
-        self.logger.debug("Got DCC Packet from queue: {command_json}".format(command_json=command))
+        self.logger.info("Got DCC Command from queue: {command_json}".format(command_json=command_json))
         command = json.loads(command_json)
```
Можно больно упасть при получении "плохого" JSON'а. Раз этот метод отвечает за разбор команды, он должен отвечать и за обработку ошибок при этом разборе с "ясным" для вызывающей стороны отчетом о ошибках (выкидывать сторого определенный набор исключений согласовынных с вызывающей стороной).

sirmax123 replied:
Fixed

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,114 @@
-                            speedDirection = {
-                                "speed": command['speed'],
-                                "direction": command['direction']}).to_bit_string(), 3)
+                speedDirection = {
```
Подобные штуки удобно делать в виде namedtuple (collections.namedtuple) - основное преимущество в этом случае орашение к полям через атрибуты `ref.speed` and `ref.direction`, плюс он не мутабельный.

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,114 @@
+    """
+    def __init__(self, commands_queue, emergency_queue,  **redis_kwargs):
+
+        self.logger = getLogger('QueueReader')
```
логгеры обычно обзывают именем класса или модуля - для имени класса можно пользовать `type(self).__name__` (но так у тебя будет светиться имя "наследника" т.е. имя именно того класса который инстанцируют, это может быть и не правильно - тогда только напрямую `QueueReader.__name__`).

sirmax123 replied:
Fixed

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,114 @@
+            self.logger.info("Command = {command_json}".format(command_json=command_json))
+            return command_json
+        else:
+            raise StopIteration
```
Эм... так же не будет работать - как только ты выкинешь `StopIteration` цикл по которому ты итерируешься прекратит свою работу... а тут задача быть бесконечным итератором... т.е. внутри должен быть цикл, который проверят данные в очередях и если их нет то засыпает на какое-то время (спать в идеале внутри операции чтения из очереди, но можно и просто `time.sleep(0.1)`).

sirmax123 replied:
Loop is moved into the queue class

sirmax123 committed:
- Refactoring: changed dict to named tuples partly
- Refactored queue classes

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,145 @@
+            self.hardware.send_bit_string(packet.to_bit_string(), 1)
 
+        except json.decoder.JSONDecodeError as Ex:
+            logger.debug(Ex)
```
Обычно логгирование в подобных местах (обработчик исключения которые перевыкидывает исключение выше) -  это плохая мысль. За обработку исключения отвечает в таких случаях кто-то другой (вышестоящий по стеку обработчик) вот он и может/должен принимать решение о необходимости логгирования или иных действий.

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,145 @@
             elif command['action'] == 'functon':
-                self.hardware.send_bit_string(self.packet_factory.DCCFunctionPacket(locoAddress = command['loco_address'],
-                                            functionsState = command['functions_state']).to_bit_string(), 3)
+                functions_state = FunctionsState(
```
тут можно сделать `functions_state = FunctionsState(**command['functions_state'])`, при условии что в дикте нет "лишних" ключей.

sirmax123 replied:
Это сложнее читать - я пока не хочу это изменять

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,145 @@
+                )
+                packet = self.packet_factory.DCCSpeedDirectionPacket(locoAddress=command['loco_address'],
+                                                                     speedDirection=speed_direction)
+                self.hardware.send_bit_string(packet.to_bit_string(), 1)
```
Этот пакет отправляется 2 раза (второй на 78 строке), это задуманное поведение?

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,145 @@
-                command_json = self.commands_queue.get()
-                self.logger.debug("Command = {command_json}".format(command_json=command_json))
-                self.decode_command(command_json)
+class RedisQueueReader(object):
```
Можно унаследоваться от `https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterator` чтобы максимально явно всем сказать что это итератор.

surabujin commented on dccpi_mm/dcc_loco_control.py:
```diff
@@ -0,0 +1,105 @@
+        }
+        self.loco_speed = 0
+        self.loco_address = loco_address
+        self.loco_direction = 'undef'
```
Почему на None?

surabujin commented on dccpi_mm/dcc_packet_factory.py:
```diff
@@ -14,28 +15,27 @@ def __init__(self, address_byte='0b1111111'):
         # Idlepacket
-        # Preamble        Address       Data           Checksum
-        #1111111111 | 0 | 11111111 | 0 | 00000000 | 0 | 11111111 1
+        # Preamb le        Address       Data           Checksum
```
случайное изменение?

surabujin commented on dccpi_mm/dcc_packet_factory.py:
```diff
@@ -122,17 +136,30 @@ def DCCFunctionPacket(self, locoAddress, functionsState = {"Fn1": 0, "Fn2": 0, "
+        # Refactoring: using namedtuple instead of dicts.
+        # Now I had to support both "old" dicts and "new"
+        # namedtuples.
+        if not isinstance(speedDirection, dict):
```
Правильнее проверять здесь на оригинальный тип `isinstance(speedDirection, SpeedDirection)`, если он не доступен, то лучшье сделать `callable(getattr(speedDirection. '_asdict', None))`

ЗЫ camelCase

sirmax123 replied:
camalCase везде в этом классе и это заплпнировано на будущий рефакторинг

surabujin commented on dccpi_mm/dcc_packet_factory.py:
```diff
@@ -215,21 +249,22 @@ def DCCDecoderControlPacket(self, locoAddress, action):
         }
 
-        if actions.has_key(action):
+        if actions in actions:
```
тут есть явно лишняя 's'

surabujin commented on dccpi_mm/dcc_redis_queue.py:
```diff
@@ -37,3 +39,44 @@ def get_nowait(self):
         return self.get(False)
 
+
+class RedisQueueReader(object):
```
Где-то я его уже видел. ;)

sirmax123 committed:
- Fix typo(s)

sirmax123 committed:
- Add extra comments
- Removed unneeded exception handler
- Removed duplicated class (moved to another file)

sirmax123 committed:
Add parent class to queue

sirmax123 committed:
Refactoring: replaced undef --> None

sirmax123 committed:
Fix emergency stop

surabujin commented on dccpi_mm/dcc_control_station.py:
```diff
@@ -4,66 +4,104 @@
+                self.hardware.send_bit_string(packet.to_bit_string(), 1)
+
+            else:
+                # If command is detected but it is not move of functiond
```
Подобная "анамалия" в протоколе, должна попадать в логи, где-то на уровне warn. Чтобы оставалось хоть какое-то понимание что что-то идет не так как задуманно.

surabujin commented on dccpi_mm/dcc_packet_factory.py:
```diff
@@ -14,28 +15,27 @@ def __init__(self, address_byte='0b1111111'):
         # Idlepacket
-        # Preamble        Address       Data           Checksum
-        #1111111111 | 0 | 11111111 | 0 | 00000000 | 0 | 11111111 1
+        # Preamble         Addres s       Data           Checksum
```
лишний пробел

sirmax123 committed:
Fix typo

sirmax123 committed:
Add warning for unexpected packets
